themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,221 @@
+date: 2022-09-30T12:28:19+02:00
+meta_desc: Pulumi is often debated to be imperative or declarative.
+  What if I tell you that Pulumi is imperative, declarative and imperative?
+meta_image: meta.png
+authors:
+    - ringo-de-smet
+tags:
+    - cloud-engineering
+    - infrastructure-as-code
+---
+
+On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
+It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?
+
+<!--more-->
+
+When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:
+
+> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
+> It’s not used for talking to the cloud provider API.
+
+It is good to know to refresh what `declarative` and `imperative` mean:
+
+> Both terms refer to how the user provides direction to the automation platform. With an imperative tool, you define the steps to execute in order to reach the desired solution. With a declarative tool, you define the desired state of the final solution, and the automation platform determines how to achieve that state. ([Source](https://www.linode.com/blog/devops/declarative-vs-imperative-in-iac/))
+
+The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
+into our product.
+
+{{< tweet id="1553431913691430912" user="adamhjk" >}}
+
+Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.
+Let me use the Pulumi architecture to highlight why I mention *imperative* twice.
+
+## Pulumi Architecture
+
+Here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:
+
+![Pulumi Architecture](/images/docs/reference/engine-block-diagram.png)
+
+You code your infrastructure in your preferred programming language. When you are done coding, you run `pulumi up` and the Pulumi CLI starts
+the language host for your selected programming language, as well as the required providers. The interaction between these 3 parts of
+the architecture results in the actual creation or modification of your infrastructure.
+
+### Language Host
+
+Under the hood, the Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in
+the `Pulumi.yaml` project file. Here is a small Python example:
+
+```python
+from pulumi_aws import s3
+
+bucket = s3.Bucket('bucket')
+
+for i in range(10):
+    s3.BucketObject(
+        f'object-{i}',
+        s3.BucketObjectArgs(
+            bucket=bucket.id,
+            key=str(i),
+        )
+    )
+```
+
+A Pulumi program models the to-be state of your infrastructure. If you read the program above, you can see that
```

Comment:
We try to avoid directional language (e.g., "above") and sensory language (e.g., "see"). https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md#language

```suggestion
A Pulumi program models the to-be state of your infrastructure. If you read this program, you can find that
```